### ISSUE TITLE
Supporting both ESM and CJS

### ISSUE THREAD

**Issue by mcollina** [giltayar/proposal-types-as-comments#108]
In order to resolve the types correctly, the type checker must know if the module is CJS, TypeScript compiled to CJS (with and without `esModuleInterop`), ESM, or faux-esm (webpack etc). 
To support that code at runtime, we need to export a module using the "infamous triplet": https://github.com/fastify/fastify/blob/main/fastify.js#L714-L727. This also enable consumers to consume the types using both CJS, ESM or TS with different configuration.

This proposal ignores the compatibility problems that are present in the ecosystem and it narrow-down to a "simple" world of ESM. I think it should be called out explicitly that the new types-as-comments are only available for modules and not for scripts. I _think_ that would be enough to resolve the problem.

@fox1t has significantly more experience on the topic to articulate this better than me.

**Comment by benjamingr** [comment]
I'm not sure I understand what the module system interop problem here is for CJS? This proposal only reserves space for types and ignores them in runtime. Doing type resolution across module system would still be a concern for the type checker anyway?

**Comment by mcollina** [comment]
> Doing type resolution across module system would still be a concern for the type checker anyway?

The type checker would still need to resolve the types across modules and packages. Right now the settings on one package would make their types incompatible with the settings I use on my modules due to the way types are exported.
The ecosystem works right now because the types are compiled away in a combo of cjs and d.ts files.
By moving the types into the language, we are solidifying the way types are exported from a module. If limiting this to modules is out of the table, I would recommend the typescript team to comment on how module resolution would work across multiple projects with potentially different tsconfig files.

**Comment by orta** [comment]
I can't speak for the TypeScript team, but it's likely that this would still work the same as it does today - where your source .js files create corresponding .d.ts files which are generated to represent the module's shape when the files are not `.ts`.

**Comment by Jamesernator** [comment]
> By moving the types into the language, we are solidifying the way types are exported from a module.

Are you saying for this issue you are wanting `export type` or such syntax supported in the script (or in CJS's case, function body) goal? The actual resolution of actually resolving where types are found has nothing to do with how they parsed.

Like settings `module.exports = SomeType` already doesn't make sense in CJS, it's more just a hack that TypeScript provides by compiling pseudo-esm into CJS, so it can just utilize `export type` without actually having that present in CJS.

I personally think it would make sense to allow `import type` and `export type` in other goals, however CJS is still weird in that it's not a script goal, it's a function body, so do we what allow `export type` in function bodies? Or do we allow `export type` in a script goal, and treat CJS as some kind of pseudo-script goal?

**Comment by mcollina** [comment]
> Like settings module.exports = SomeType already doesn't make sense in CJS, it's more just a hack that TypeScript provides by compiling pseudo-esm into CJS, so it can just utilize export type without actually having that present in the transpiled CJS.

Thanks. This is a much better way to frame the problem. __How would this proposal "fit" within CJS__? How would we export and import a type?

**Comment by fox1t** [comment]
@mcollina, thanks for mentioning me because these were my concerns too when I read the tweets and related the proposal.

It is important to remember that a large part of CJS only module types today are written using [TS `namespaces`](https://www.typescriptlang.org/docs/handbook/namespaces.html). `namespaces` are not deprecated but are not compatible with ES Modules either. 

We need to answer this question: how will this new type system represent `module.exports = {...}` or `module.exports = function(...) {...}`? Today we don't have this answer since, as @mcollina pointed out, every current build system has its own way of handling this. I think, though, that we should agree that `export default` and `module.exports` should be two different things in this new JS type system. As @benjamingr already pointed out, the type checkers that will rise should handle this problem on their side.

 I would love us to focus on JS type system without trying to make it compatible with the current state of the build systems and type checkers.

P.S. Please let's retire the `infamous triplet` with this proposal! Even if I somehow discovered it and evangelized its use as the method compatible with almost every build system configuration, I don't like it. It is there to remind us that we need to make things better.

**Comment by Jamesernator** [comment]
> It is important to remember that a large part of CJS only module types today are written using [TS `namespaces`](https://www.typescriptlang.org/docs/handbook/namespaces.html). `namespaces` are not deprecated but are not compatible with ES Modules either.

This is well beyond the scope of the proposal here, but personally I rather like the namespaces feature for implementing, well namespace-esque things similar to `JSON`. I wonder if it would be worth having another proposal for `namespace` as an actual JS feature that is roughly a immediately intialized module.

**Comment by giltayar** [comment]
@mcollina I'm confused. You said:

> *The type checker* would still need to resolve the types across modules and packages. Right now the settings on one package would make their types incompatible with the settings I use on my modules due to the way types are exported.

And I agree. But the type checking is not done by the JavaScript runtime, but by external 3rd party tooling. the JS runtime just ignores the types. So the ins and outs of CJS vs ESM in type checking is not relevant to this proposal?

**Comment by mcollina** [comment]
This proposal completely ignore the split between CJS and ESM. I don't think it should... it should be explicitly calling out the problems of CJS vs TS vs ESM and calling out the import/export problem.... deciding to do _nothing_ about it, essentially cutting CJS out.

**Comment by benjamingr** [comment]
@giltayar I think Matteo's point if I understand it is a bit like:

 - Node has two module systems 
 - Currently people using TypeScript almost always write ESM and transpile to CJS _or_ use ESM directly.
 - This proposal adds type comment space to JavaScript, this adds syntax that currently don't need ESM in TypeScript because of transpilation like `import type`. This won't work in CJS and there are no alternatives/recourse in the language.


I think this is fine probably? ECMAScript only specifies ES Modules and not CJS so it makes sense for it to reserve `import type` - I think reserving syntax to do the same thing in CJS is up to Node and TypeScript to figure out probably and is out of scope?

I think in any case it would make sense to mention this issue and considerations in the README

**Comment by mcollina** [comment]
Thanks @benjamingr! Acknowledging the problem and noting down that _types as comments won't work for CJS_ is enough.

**Comment by fox1t** [comment]
I agree that it should be written down explicitly. I want to add that the only blocker for "types in CJS" is the import/export part. 
```ts
// foo.cjs
type FooReturnValue { 
   foo: string | number 
}

module.exports = function foo(arg1: string, arg2?: number): FooReturnValue {
   return {
    foo: arg2 ?? arg1
  }
}
```
```ts
// bar.cjs
const foo = require('./foo.cjs')

const myFoo = foo('hi', 3) // [1]
```
`[1]`:  Even if the type checker (TS? Flow? A new one?) figures out that `myFoo` has `FooReturnValue` type, there is no way to export `FooReturnValue` from `foo.cjs`. It is mainly a Node.js problem, but the spec should cite it since CJS is also widely used in other contexts.

I think that out of the box, there will be support for types in CJS (since, as @ljharb already said, the parser is the same and CJS is valid `JS` indeed), but I don't see any feasible way that allows us to export types from the file itself. 

As @orta said, the way to go should probably be a `.d.cjs` file (I would like to avoid the current `.ts` extension if possible) that uses ESM syntax and the new Types as comment definitions to expose the public API of the file. And that can be picked by type checkers.

**Comment by Jamesernator** [comment]
> I think reserving syntax to do the same thing in CJS is up to Node

If Node allowed some form of type import/export syntax within CJS, wouldn't it then have the preprocess every CJS file using a runtime parser rather than just handing it off to v8? Like if we actually specify type exports/imports for the function body goal, then the parsing can entirely be handed off to v8.

> As @orta said, the way to go should probably be a `.d.cjs` file (I would like to avoid the current `.ts` extension if possible) that uses ESM syntax and the new Types as comment definitions to expose the public API of the file. And that can be picked by type checkers.

The current extensions in TypeScript's proper Node ESM support just mirrors whatever extension is used, i.e. if you have CJS in a `.cjs` file then you use `.d.cts`, if it's in a `.js` file (i.e. not using `"type": "module"` in `package.json`, the status quo) then you can just use `.d.ts` per usual.

**Comment by Jamesernator** [comment]
Actually I wonder if another way to allow CJS to support importing/exporting types correctly would be for the proposal to support `declare` (currently this is [undecided](https://github.com/giltayar/proposal-types-as-comments#ambient-declarations)).

If we allowed `declare`, type systems could simply define a new declarable thing (e.g. like how `declare global` is special, we'd add a new declare kind) to perform these exports:

```ts
// myFile.cjs
declare types {
    import type Foo from "./Foo.cjs";
    
    export type Bar = { foo: Foo };
}
```

**Comment by fox1t** [comment]
> The current extensions in TypeScript's proper Node ESM support just mirrors whatever extension is used, i.e. if you have CJS in a `.cjs` file then you use `.d.cts`, if it's in a `.js` file (i.e. not using `"type": "module"` in `package.json`, the status quo) then you can just use `.d.ts` per usual.

Sure, but what I meant is that I would like to add to this specification the part that allows developers to write external definition files that are ignored by the JS runtimes but can be picked up by type checkers. 

> Actually I wonder if another way to allow CJS to support importing/exporting types correctly would be for the proposal to support `declare` (currently this is [undecided](https://github.com/giltayar/proposal-types-as-comments#ambient-declarations)).
> 
> If we allowed `declare`, type systems could simply define a new declarable thing (e.g. like how `declare global` is special, we'd add a new declare kind) to perform these exports:
> 
> ```ts
> // myFile.cjs
> declare types {
>     import type { Foo } from "./Foo.cjs";
>     
>     export type Bar = { foo: Foo };
> }
> 
> const foo = require("./Foo.cjs");
> cjsCodeHereAsUsual();
> ```
I really like this! It allows "inlining" the even in CJS!

**Comment by mcollina** [comment]
I think `declare` would work @Jamesernator.

**Comment by acutmore** [comment]
From my experience importing types (into scripts, using TypeScript) has been less of an issue (compared to exporting).

I would use something like: `type x = (typeof import('file.d.ts'));`

**Comment by giltayar** [comment]
@mcollina @benjamingr isn't this problem related _only_ to `import type` and to the whole reserved syntax space? If so, the final proposal should figure out what to do with `import type` in "script" code: should it be allowed or disallowed. This is a good question! We're at stage 0, so questions like this need to be raised, but not necessarily answered.

**Comment by mcollina** [comment]
The same is true for `export type`. I agree this problem just be solved before it reaches stage 3/4, totally not now.

**Comment by fox1t** [comment]
> @mcollina @benjamingr isn't this problem related _only_ to `import type` and to the whole reserved syntax space? If so, the final proposal should figure out what to do with `import type` in "script" code: should it be allowed or disallowed. This is a good question! We're at stage 0, so questions like this need to be raised, but not necessarily answered.
> 
> 
> 
> 

I agree we are on time! Still, I really like the proposal of using the “declare scope” to allow exporting/importing types in CJS.
